### ISSUE TITLE
Move config from Startup to Config

### ISSUE THREAD

**Issue by grokky1** [NLog/NLog.Web#172]
**TL;DR**
Below are some changes needed to make most projects work with ASP.NET Core2 (netcoreapp2.0). I've also submitted PRs for these changes.

---

There are a number of "please upgrade NLog to ASP.NET Core 2.0" issues, but this is a little different.

There are a few workarounds documented here to make NLog work on netcoreapp2.0. I have it working.

However, logging is now split over `Startup.cs` and `Program.cs`. In 1.1 all logging config was done in `Startup.Configure()`. But in 2.0 logging is configured as soon as the app starts, in `Program.BuildWebHost(string[])`. That makes logging available immediately, including from within the startup code.

Right now default logging is configured in `Program`, and NLog is configured later in `Startup.Configure()`. It should be moved into `Program`.

Here is an example of something that doesn't work because of this. I do this to log app lifetime events:

````csharp
public Startup(IConfiguration configuration, ILogger<Startup> logger)
{
  _configuration = configuration;
  _logger = logger;
}

private readonly IConfiguration _configuration;
private readonly ILogger<Startup> _logger;

public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory, IApplicationLifetime appLifetime)
{
  // other stuff...
  appLifetime.ApplicationStarted.Register(Started);
  appLifetime.ApplicationStopping.Register(Stopping);
  appLifetime.ApplicationStopped.Register(Stopped);
  // other stuff...
  // configure NLog in the usual way
}

private void Started()
{
  _logger.LogInformation("App started");
}

private void Stopping()
{
  _logger.LogInformation("App stopping");
}

private void Stopped()
{
  _logger.LogInformation("App stopped");
}
````

The `Started` event is called, and it logs to the Console and Debugger loggers. But not to NLog. That's because NLog was configured too late.

To make it work - instead of injecting `ILogger<Startup>` into the constructor, I can do this in `Configure()` after setting up NLog: `_logger = loggerFactory.CreateLogger<Startup>();`.

As you can see this is messy. All config must be in `Program`.

This can be hotfixed easily. In the `Program.BuildWebHost(string[])` we could add this:

````csharp
return WebHost
  .CreateDefaultBuilder(args)
  .ConfigureLogging((context, builder) => {
    context.HostingEnvironment.ConfigureNLog("nlog.config");
    builder.AddNLog();         // requires FIX 1
    builder.AddNLogWeb();      // requires FIX 2
  })
  .UseStartup<Startup>()
  .Build();
````

To make that work:

## FIX 1
Must use @xperiandri's [code found here](https://github.com/NLog/NLog.Extensions.Logging/issues/108#issuecomment-323119458)

## FIX2
Must [change this](https://github.com/NLog/NLog.Web/blob/master/NLog.Web.AspNetCore/AspNetExtensions.cs#L26)

````csharp
public static void AddNLogWeb(this IApplicationBuilder app)
{
  ServiceLocator.ServiceProvider = app.ApplicationServices;
}
````

to this:

````csharp
public static void AddNLogWeb(this ILoggingBuilder loggingBuilder)
{
  ServiceLocator.ServiceProvider = loggingBuilder.Services;
}
````

Once these changes are made, it should work out of the box for netcoreapp2.0. There are some config options for xml files, etc, which I guess someone else should look at as I don't use that.

I've submitted PRs to both repos for [fix1](https://github.com/NLog/NLog.Extensions.Logging/pull/126) and [fix2](https://github.com/NLog/NLog.Web/pull/173). Maybe @xperiandri or @304NotModified could have a look?
